File: Top-Interview-Questions/Array-and-Strings-Medium-Collection/334-increasing-triplet-subsequence.js
/**
 * @param {number[]} nums
 * @return {boolean}
 * Runtime: 96 ms, faster than 22.32% of JavaScript online submissions for Increasing Triplet Subsequence.
 * Memory Usage: 36 MB, less than 35.97% of JavaScript online submissions for Increasing Triplet Subsequence.
 */
var increasingTriplet = function(nums) {
    if (nums === null || nums.length<3 ) {
        return false;
    }

    let small = nums[0], mid = Infinity;

    for (let i=1; i<nums.length; i++) {
        if (nums[i] > small && nums[i] < mid) {
            mid=nums[i];
        }

        if (nums[i] < small) {
            small=nums[i];
        }

        if (nums[i] > mid) {
            return true;
        }
    }

    return false;
};

File: Top-Interview-Questions/Array-and-Strings-Medium-Collection/334-increasing-triplet-subsequence.py
class Solution:
    def increasingTriplet(self, nums: List[int]) -> bool:
        if not nums or len(nums) < 3:
            return False
        
        second = float('inf')
        third = float('inf')
        for first in nums:
            if first <= second:
                second = first
            elif first <= third:
                third = first
            else:
                return True
            
        return False
        # TC: O(n)
        # SC: O(1)
        # Runtime: 80 ms, faster than 26.59% 
        # Memory Usage: 14.6 MB, less than 17.36% 